BUG: Preserve identity of dtypes in make_mask_descr

eric-wieser opened the pull request:
Partially addresses #8666

eric-wieser committed:
BUG: Preserve identity of dtypes in make_mask_descr

Partially addresses #8666

mhvk commented on numpy/ma/core.py:
```diff
@@ -1296,14 +1296,23 @@ def _recursive_make_descr(datatype, newtype=bool_):
                 name = (field[-1], name)
             descr.append((name, _recursive_make_descr(field[0], newtype)))
-        return descr
+        newtype = np.dtype(descr)
```
Maybe instead of calling this `newtype` (which is confusing as it is also a passed-in argument), just keep calling it `descr`?

mhvk commented on numpy/ma/core.py:
```diff
@@ -1296,14 +1296,23 @@ def _recursive_make_descr(datatype, newtype=bool_):
         mdescr = list(datatype.subdtype)
         mdescr[0] = _recursive_make_descr(datatype.subdtype[0], newtype)
-        return tuple(mdescr)
+        newtype = np.dtype(tuple(mdescr))
```
idem

mhvk commented on numpy/ma/core.py:
```diff
@@ -1296,14 +1296,23 @@ def _recursive_make_descr(datatype, newtype=bool_):
+
+    # promote the type to a dtype if otherwise untouched
     else:
-        return newtype
```
I think for this private function, we should be able to count on `newtype` being a `dtype` already, so I would leave the `return newtype` here. (I realise that means updating the two occurrences where `"O"` is passed in.)

eric-wieser replied:
Every top-level call-site in the file fails to pass in a `dtype` right now (either `MaskType`, `'O'`, or `bool`, I think), so this is absolutely required. Sure, we could change all the callsites, I guess...

mhvk replied:
I think that would be better, but it is a very mild preference at best, so also fine to leave it. For easy following of the flow of logic, I would still think it is best to not reuse the `newtype` name, though (it really threw me off, to convince me that the recursive calls could not pick that up).

eric-wieser replied:
If we do that, then I'd be tempted to change `MaskType` from `np.bool_` to `np.dtype(np.bool_)`

mhvk reviewed (approved):
Looks good, with some nitpicks that really are a matter of taste, so fee free to ignore.

mhvk commented:
My feeling would be to have this for 1.13 rather than 1.12.1. What do you think?

mhvk commented on numpy/ma/core.py:
```diff
@@ -1296,14 +1296,23 @@ def _recursive_make_descr(datatype, newtype=bool_):
+        newtype = np.dtype(newtype)
+
+    # preserve identity of dtypes
+    if newtype == datatype:
```
With the above, the `np.dtype` calls can be here, i.e.,
```
descr = np.dtype(descr)
if descr == datatype:
    return datatype
else:
    return descr
```

eric-wieser commented:
Yeah, I agree - no pressure for 12.1 here

eric-wieser commented:
@mhvk : Made the changes sort of like you requested

eric-wieser commented on numpy/ma/core.py:
```diff
@@ -1285,25 +1285,48 @@ def __str__(self):
 
-def _recursive_make_descr(datatype, newtype=bool_):
-    "Private function allowing recursion in make_descr."
+def _replace_dtype_fields_recursive(dtype, primitive_dtype):
```
`_make_descr` was a pretty bad name

eric-wieser commented on numpy/ma/core.py:
```diff
@@ -1285,25 +1285,48 @@ def __str__(self):
+    """
+    dtype = np.dtype(dtype)
+    primitive_dtype = np.dtype(primitive_dtype)
+    return _replace_dtype_fields_recursive(dtype, primitive_dtype)
```
Now do the dtype conversions just once, so we don't need to again

eric-wieser commented on numpy/ma/core.py:
```diff
@@ -1285,25 +1285,48 @@ def __str__(self):
+    return new_dtype
+
+
+def _replace_dtype_fields(dtype, primitive_dtype):
```
Possible scope for public visibility? I can see "convert all fields to object/bool" being marginally desirable elsewhere

eric-wieser commented on numpy/ma/core.py:
```diff
@@ -3834,7 +3854,7 @@ def __str__(self):
                     res.view(ndarray)[mask] = f
                 else:
-                    rdtype = _recursive_make_descr(self.dtype, "O")
+                    rdtype = _replace_dtype_fields(self.dtype, "O")
```
Implementation details of `mask_make_descr` shouldn't be used elsewhere!

eric-wieser commented on numpy/ma/core.py:
```diff
@@ -1285,25 +1285,48 @@ def __str__(self):
+    return _replace_dtype_fields_recursive(dtype, primitive_dtype)
 
 
 def make_mask_descr(ndtype):
```
Ideally I'd rename this to `make_mask_dtype` or `mask_dtype_for`, but that's for another PR.

mhvk commented:
I like this very much! My only comment is that with this change it is as much `MAINT` as it is `BUG`...

eric-wieser commented:
Arguably `BUG` should be `ENH` here anyway, since identity comparison of dtypes is not really promised anywhere

ahaldane commented on numpy/ma/core.py:
```diff
@@ -1285,25 +1285,48 @@ def __str__(self):
                 name = (field[-1], name)
-            descr.append((name, _recursive_make_descr(field[0], newtype)))
-        return descr
+            descr.append((name, _replace_dtype_fields_recursive(field[0], primitive_dtype)))
```
line is too long

ahaldane commented:
So if I understand, the main bug fix is the test for `if new_dtype == dtype:  new_dtype = dtype` and the other changes are for style? That's fine, I just want to understand the motivation for the other changes.

eric-wieser commented:
@ahaldane: Yep, that's correct. I'll rewrite this into two commits to emphasize that

eric-wieser committed:
MAINT: Don't call internal recursion helpers elsewhere

Cleans up make_mask_descr, and adds a private _replace_dtype_fields, which
we need elsewhere to replace all dtypes with `object` instead of `bool`.

This new version also removes repeated calls to `np.dtype`, doing the conversion
only once.

eric-wieser commented:
@ahaldane: Rebased, and long lines shortened. First commit is enough to make the tests pass

eric-wieser commented on numpy/ma/core.py:
```diff
@@ -1334,13 +1359,10 @@ def make_mask_descr(ndtype):
     dtype([('foo', '|b1'), ('bar', '|b1')])
     >>> ma.make_mask_descr(np.float32)
-    <type 'numpy.bool_'>
+    dtype('bool')
```
This was already the behaviour before the PR - the docs were just wrong

mhvk commented:
I'm now even happier with this, so @ahaldane, if you agree as well, I think this can just be merged.

ahaldane commented:
Looks good, I'll go ahead with the merge then. Thanks @eric-wieser !
